app/src/org/commcare/utils/EncryptionUtils.java

```diff
@@ -1,17 +1,212 @@
 package org.commcare.utils;
 
+import android.os.Build;
+import android.security.keystore.KeyGenParameterSpec;
+import android.security.keystore.KeyProperties;
+
+import androidx.annotation.RequiresApi;
+
 import org.commcare.util.Base64;
+import org.commcare.util.Base64DecoderException;
 
+import java.io.IOException;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
 import java.security.MessageDigest;
 import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.UnrecoverableEntryException;
+import java.security.cert.CertificateException;
+import java.util.Random;
+
+import javax.crypto.BadPaddingException;
+import javax.crypto.Cipher;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.KeyGenerator;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.IvParameterSpec;
 
 /**
  * Utility class for encrypting submissions during the SaveToDiskTask.
  *
  * @author mitchellsundt@gmail.com
  */
+
 public class EncryptionUtils {
 
+    private static final int PASSPHRASE_LENGTH = 32;
+    private static final String SECRET_NAME = "secret";
+
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    private static final String ALGORITHM = KeyProperties.KEY_ALGORITHM_AES;
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    private static final String BLOCK_MODE = KeyProperties.BLOCK_MODE_CBC;
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    private static final String PADDING = KeyProperties.ENCRYPTION_PADDING_PKCS7;
+    private static KeyStore keystore = null;
+
+    //Gets the SecretKey from the Android KeyStore (creates a new one the first time)
+    private static SecretKey getKey()
+            throws CertificateException, KeyStoreException, IOException, NoSuchAlgorithmException,
+            UnrecoverableEntryException, InvalidAlgorithmParameterException {
+        if(keystore == null) {
+            keystore = KeyStore.getInstance("AndroidKeyStore");
+            keystore.load(null);
+        }
+
+        KeyStore.Entry existingKey = keystore.getEntry(SECRET_NAME, null);
+        if (existingKey instanceof KeyStore.SecretKeyEntry) {
+            return ((KeyStore.SecretKeyEntry) existingKey).getSecretKey();
+        }
+
+        //Create key
+        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
```

Comment:
if we remove this check, what exactly fails on devices less than Android M ?

Reply:
Everything inside the if-block requires at least Android M , starting with the `KeyGenParameterSpec.Builder()` constructor, including the 4 setters called on the `builder`, and the final call to `builder.build()`.

The two (private) callers for this method both require Android M, so depending on how we want older versions of Android to handle this we should make this method consistent with the two callers.

Reply:
This code was replaced in the implementation that supports Android before version M.